src/yachtzee/fishy.ts

```diff
@@ -0,0 +1,203 @@
+import {
+  availableAmount,
+  booleanModifier,
+  canAdventure,
+  cliExecute,
+  equip,
+  haveEffect,
+  haveEquipped,
+  mallPrice,
+  print,
+  use,
+} from "kolmafia";
+import {
+  $effect,
+  $item,
+  $location,
+  adventureMacro,
+  get,
+  getActiveEffects,
+  have,
+  Macro,
+  sum,
+  uneffect,
+} from "libram";
+import { acquire } from "../acquire";
+import { safeRestore } from "../lib";
+import { pyecAvailable, yachtzeeBuffValue } from "./lib";
+import { getBestWaterBreathingEquipment } from "./outfit";
+
+
+function fishyCloverAdventureOpportunityCost(pipe: boolean) {
+  const willBeFishy = pipe || have($effect`Fishy`);
+  const fishyCloverAdventureCost = willBeFishy ? 1 : 2;
+  const adventureExtensionBonus = pyecAvailable() ? 5 : 0;
+  return sum(getActiveEffects(), (currentBuff) => {
+    const buffValue = yachtzeeBuffValue(currentBuff);
+    if (buffValue <= 0) return 0;
+
+    const currentBuffTurns = haveEffect(currentBuff);
+    if (currentBuffTurns <= fishyCloverAdventureCost) {
+      return (currentBuffTurns + adventureExtensionBonus) * buffValue;
+    }
+    return fishyCloverAdventureCost * buffValue;
+  });
+}
+
+export function optimizeForFishy(yachtzeeTurns: number, setup?: boolean): number {
+  // Returns the lowest cost for fishy
+  // Assume we already maximized for meat; this returns the cost of swapping out meat% equips for underwater breathing equips
+  const bestWaterBreathingEquipment = getBestWaterBreathingEquipment(yachtzeeTurns);
+
+  if (
+    setup &&
+    !have($effect`Really Deep Breath`) &&
+    bestWaterBreathingEquipment.item !== $item.none
+  ) {
+    equip(bestWaterBreathingEquipment.item);
+  }
+  if (
+    haveEquipped($item`The Crown of Ed the Undying`) &&
+    !booleanModifier("Adventure Underwater")
+  ) {
+    cliExecute("edpiece fish");
+  }
+  // If we already have fishy, then we longer need to consider the cost of obtaining it
+  if (haveEffect($effect`Fishy`) >= yachtzeeTurns) return 0;
+
+  // Restore here if we potentially need to visit an adventure.php zone to grab fishy turns
+  if (haveEffect($effect`Beaten Up`)) {
+    uneffect($effect`Beaten Up`);
+  }
+  safeRestore();
+
+  const haveFishyPipe = have($item`fishy pipe`) && !get("_fishyPipeUsed");
+  const adventureExtensionBonus = pyecAvailable() ? 5 : 0;
+  const fishySources = [
```

Comment:
I've actually thought about changing this structure slightly - I'd like a field called `baseFishyTurns` or `fishyTurns` (with/out consideration of PYEC) so that we don't need to change the cost to infinity, which is (1) ugly, and (2) prone to making stupid mistakes like I did at the start when adding pure fishy pipe support for shorter chains

Reply:
Also, as acco has noted in passing, when we are considering shorter chains (e.g. base 10 or 20), we should also consider pocket wishes as a fishy source - they may very well be optimal

Reply:
Out of scope of this PR